src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/NativeProjectSystemReferencesReader.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using EnvDTE;
+using Microsoft;
+using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.VCProjectEngine;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.ProjectManagement;
+using NuGet.ProjectModel;
+using NuGet.VisualStudio;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Reference reader implementation for the core project system in the integrated development environment (IDE).
+    /// </summary>
+    internal class NativeProjectSystemReferencesReader
+        : IProjectSystemReferencesReader
+    {
+        private readonly IVsProjectAdapter _vsProjectAdapter;
+        private readonly IVsProjectThreadingService _threadingService;
+        private readonly AsyncLazy<VCProject> _vcProject;
+
+        public NativeProjectSystemReferencesReader(
+            IVsProjectAdapter vsProjectAdapter,
+            IVsProjectThreadingService threadingService)
+        {
+            Assumes.Present(vsProjectAdapter);
+            Assumes.Present(threadingService);
+
+            _vsProjectAdapter = vsProjectAdapter;
+            _threadingService = threadingService;
+            _vcProject = new AsyncLazy<VCProject>(async () =>
+            {
+                await threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
+                return _vsProjectAdapter.Project.Object as VCProject;
+            }, threadingService.JoinableTaskFactory);
+        }
+
+        public async Task<IEnumerable<ProjectRestoreReference>> GetProjectReferencesAsync(
+            Common.ILogger logger, CancellationToken _)
+        {
+            var results = new List<ProjectRestoreReference>();
+
+            await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
+            var vcProject = await _vcProject.GetValueAsync();
+            var references = vcProject.VCReferences as VCReferences;
+            var projectReferences = references.GetReferencesOfType((uint)vcRefType.VCRT_PROJECT);
+
+            foreach (var reference in projectReferences)
+            {
+                var vcReference = reference as VCProjectReference;
+                if (vcReference.UseInBuild)
+                {
+                    var childProjectPath = (vcReference.ReferencedProject as Project).FileName; // TODO: This returns the assembly path, but we really want the vcxproj path instead.
```

Comment:
ReferencedProject can return null in some cases, so best check for that.
Also, the TODO comment is now outdated, right?

Reply:
Yeah the TODO comment is outdated. 

What would it mean if a Project is null? 
I do search for project references only, so just want to make sure if we need to try to use some fallback to get the project path?

What would happen if the project reference is not loaded?

Reply:
It looks to me like in extreme cases it can be null like in shutdown and or project unload cases which isn't that extreme. 
Unfortunately, it looks as if you'd need to do some extra work if the reference is to an unloaded project. Such as going to the IVsSolution and looking up via VCProjectReference.ReferencedProjectIdentifier with GetProjectOfGuid if the identifier can be converted to Guid via that identifier string, or as a fallback GetProjectOfUniqueName. Then, the IVsHierarchy you get back should be to the stub object and then IVsHierarchy.GetCanonicalName(VSITEMID_ROOT...) would give you the project filename. Note that I think you'd need to handle the case where even after all of that, you still can't find the information (like let's say the reference project was removed), so what else could you do other than skip it as if it didn't exist?

Reply:
Knowing about unloaded projects enables Solution Filters to work with unloaded projects. 

Example: 

P1 -> P2 -> P3, knowing about the exact reference even when P2 is unloaded allows the builds work just work.

That isn't concern with native projects though, so we can just ignore it.